### PR TITLE
Oauth2 filter: refresh token support

### PR DESCRIPTION
**Pull request by Alexcei88**
Commit Message:
Using a refresh token for updating the access token if that is provided by identity provider server
Additional Description:

A refresh token is a special token that is used to obtain additional access tokens. This allows you to have short-lived access tokens without having to collect credentials every time one expires.  If a refresh token is provided by oauth endpoint then after that the access token is expired the filter gets a new access token using by a refresh token without having to ask the user to log in again.
I see these changes to resolve for [this ](https://github.com/envoyproxy/envoy/issues/23007) problem

For backward compatilibility the special flag `use_refresh_token `was added to configuration. By default it is false, so the refresh token is not used. Its old behaviour. To turn on using refresh token it's needed to manually add `use_refresh_token ` to configuration of the filter.

To decrease duplication of code in the filter some of the pieces of code were extracted to separate functions in the filter. Therefore it looks like a lot of changes but it's not. I just refactored code to do it beatiful.

Also I added additional statistic counters for filter about using refresh token.

Risk Level:
Medium 

Testing:

Need to test oauth2 filter

Docs Changes:

Need to add description about new flag in the configuration

Release Notes:

Need to mention about supporting of refresh token in the filter

### CODE REVIEW

**Comment by KBaichoo** [comment]
/assign @derekargueta 

Can you take a look as oauth2 filter owner? Thank you.

/wait

For CI fixes

**Comment by Alexcei88** [comment]
##[error]No space left on device

I have already catched this exception two times today. How to restart CI?

**Comment by Alexcei88** [comment]
It looks like CI is fixed

**Comment by KBaichoo** [comment]
/assign @snowp 

As other oauth owner.

I think @derekargueta might not be active.

**Comment by Alexcei88** [comment]
Hi @snowp , thank you for taking a look.
I would say, it is impossible. If we reduce refactoring then we will duplicate functionality. Then there will be more changes than now. 

I understand that it’s hard to see in commits, but I didn’t touch the old functionality at all. I just added a new one, but it is a quite large. New branch of logic was added to filter.

But I am sure for changes, because it already works on my production environment.

One moment I need to note, which I don't like in my implementation there is no checking a refresh token livespan. So sometimes we try to update access token using by expired refresh token. In this case the attempt finish as failed and user will be redirect to log in again. In the future we can add validation of refresh token before attempting update access token by refresh token.

**Comment by snowp** [comment]
Unfortunately it'll be hard to carve out the time to review this as it stands, I'd suggest first doing a PR that performs all the refactoring that this PR does to the existing code, and then once that has been reviewed do another one which contains the addition of the refresh token support. Does that make sense?

**Comment by Alexcei88** [comment]
@snowp, Your suggestion sounds good. 
But I think I can't spend more time on this feature. 
I think I made a hard work to add this cool feature in the filter. 
If I will have time to  separate PR  I make it. But I am no sure because it will required additional time.

**Comment by Alexcei88** [comment]
@snowp , I submitted otheer PR as you requested
https://github.com/envoyproxy/envoy/pull/25160

**Comment by wbpcode** [comment]
@Alexcei88 could you merge main and resolve the conflict? Thanks.

**Comment by snowp** [comment]
There seems to be no commits left in this PR, did something go wrong during the merge?

**Comment by Alexcei88** [comment]
Thank you for your attention. Yes, I cleared all changes in the branch. I will soon commit significant changes in the branch. Wait me please.

**Comment by Alexcei88** [comment]
ready for review

File: api/envoy/extensions/filters/http/oauth2/v3/oauth.proto
Comment (lizan) [inline api/envoy/extensions/filters/http/oauth2/v3/oauth.proto:119]: Should this default to true? If so use BoolValue and document that.

File: source/extensions/filters/http/oauth2/filter.cc
Comment (lizan) [inline source/extensions/filters/http/oauth2/filter.cc:582]: no need `.reset()`.

File: source/extensions/filters/http/oauth2/filter.cc
Comment (lizan) [inline source/extensions/filters/http/oauth2/filter.cc:574]: `absl::StrJoin(cookies, "; ", absl::PairFormatter("="));`

File: source/extensions/filters/http/oauth2/filter.cc
Comment (lizan) [inline source/extensions/filters/http/oauth2/filter.cc:551]: I don't think you need set those cookies in request to response_headers_to_add again, just new cookies should be fine.

Comment (Alexcei88) [reply source/extensions/filters/http/oauth2/filter.cc:551]: You suggest to not save new cookies to the response_headers_to_add?
I tried it before, but it doesn't works. I didn't deep why it is. So I just decided to save cookies and set cookies through set-cookie headers in the response

File: api/envoy/extensions/filters/http/oauth2/v3/oauth.proto
Comment (Alexcei88) [reply api/envoy/extensions/filters/http/oauth2/v3/oauth.proto:119]: The default is false. Because I don't consider to change old behaviour after update envoy. If user wants to use refresh token then manually set the flag to true.

File: source/extensions/filters/http/oauth2/filter.cc
Comment (lizan) [reply source/extensions/filters/http/oauth2/filter.cc:551]: I'm suggesting not set all `Cookie` in request header to `Set-Cookie` as this might cause a security issue that a new `Set-Cookie` can overwrite the original expiration, or remove some property like `Secure` or `HttpOnly`.

Comment (Alexcei88) [reply source/extensions/filters/http/oauth2/filter.cc:551]: `Set-Cookie` is applied only for specific cookies, related with OAuth2: id Token, RefreshToken, AccessToken, OAuth Hmac, OAuth Expires. Other cookies which recieved in the request I resend in initial state. I cannot add additional header `Cookie` in the request. As I know, only one header `Cookie` in the request is allowed. Looks like I don't catch your suggestion, I am sorry. 
Could you explain again please what I need to do?

**Comment by jmarantz** [comment]
@Alexcei88 please do not force-push in Envoy pull requests.

See https://github.com/envoyproxy/envoy/blob/main/CONTRIBUTING.md

> Once your PR is under review, please do not rebase it. If you rebase, you will need to force push to github, and github's user interface will force your reviewer to review the PR from scratch rather than simply look at your latest changes.  ...

**Comment by Alexcei88** [comment]
Could anyone have a chance to look?

**Comment by snowp** [comment]
Yeah I will take a look sometime today, sorry for all the delay here

File: source/extensions/filters/http/oauth2/filter.cc
Comment (snowp) [reply source/extensions/filters/http/oauth2/filter.cc:551]: My understanding (and maybe @lizan can clarify if I'm misrepresenting what he's saying) is that you don't need to include all the cookies that are present on the request headers in the response headers, only the ones that you are trying to update. That way you are not resetting more cookies than necessary

File: test/extensions/filters/http/oauth2/filter_test.cc
Comment (snowp) [inline test/extensions/filters/http/oauth2/filter_test.cc:1368]: remove this newline

File: test/extensions/filters/http/oauth2/filter_test.cc
Comment (snowp) [inline test/extensions/filters/http/oauth2/filter_test.cc:1425]: Since we don't check state_ i don't see much reason mentioning it

File: test/extensions/filters/http/oauth2/filter_test.cc
Comment (snowp) [inline test/extensions/filters/http/oauth2/filter_test.cc:1367]: Maybe add a description to this test talking about what this tests - seems like it first does one authorization that doesn't use the refresh token followed by one that does?

File: test/extensions/filters/http/oauth2/filter_test.cc
Comment (snowp) [inline test/extensions/filters/http/oauth2/filter_test.cc:1645]: This doesn't seem like the third request, update?

**Review by snowp — DISMISSED** [review]
Thanks, the overall flow seems right to me, but this needs a bit of work otherwise

I think this would benefit from having a test in oauth_integration_test that verifies that we correctly pause filter iteration while waiting for a response for the refresh request. There are also a few branches that need test coverage per the coverage report (for example in https://storage.googleapis.com/envoy-pr/e9fdd37/coverage/source/extensions/filters/http/oauth2/filter.cc.gcov.html )

File: source/extensions/filters/http/oauth2/oauth_client.cc
Comment (Alexcei88) [reply source/extensions/filters/http/oauth2/oauth_client.cc:139]: NetworkErrorDoubleCallStateInvalid test covers this exception

File: source/extensions/filters/http/oauth2/filter.cc
Comment (Alexcei88) [reply source/extensions/filters/http/oauth2/filter.cc:568]: Yep, we need to update BearerToken, OauthHMAC, OauthExpires  to send them in the actual state to the upstream.

File: source/extensions/filters/http/oauth2/oauth_client.cc
Comment (Alexcei88) [reply source/extensions/filters/http/oauth2/oauth_client.cc:178]: This exception is not covered by any tests

**Comment by Alexcei88** [comment]
@jmarantz thank you for information. But I use a force push only for satisfying DCO requirement if I forget to sign my  commit

File: api/envoy/extensions/filters/http/oauth2/v3/oauth.proto
Comment (dio) [inline api/envoy/extensions/filters/http/oauth2/v3/oauth.proto:128]: Probably update this description to something like: 

"Set to true to enable automatic access token refresh using the associated refresh token, provided that the OAuth server supports this feature."

File: changelogs/current.yaml
Comment (dio) [inline changelogs/current.yaml:159]: This is probably a bad merge (?).

File: source/extensions/filters/http/oauth2/filter.cc
Comment (dio) [inline source/extensions/filters/http/oauth2/filter.cc:376]: Sorry. Could you add a comment on why we need this condition? Thank you.

File: source/extensions/filters/http/oauth2/filter.cc
Comment (dio) [inline source/extensions/filters/http/oauth2/filter.cc:515]: Probably `onRefreshAccessTokenSuccess` is a better name (?).

File: source/extensions/filters/http/oauth2/oauth_client.cc
Comment (dio) [reply source/extensions/filters/http/oauth2/oauth_client.cc:139]: Probably we can explain more about when this case will happen in a real case (?). Also when this crashes envoy, will that be because of the response from the server or upstream?

**Review by dio — COMMENTED** [review]
Thank you @Alexcei88. Flushing some comments. Also probably we need an example of using this filter under the: `examples` directory (this will be in a separate PR though, seems like you have it on one of your branches: `oauth2_refresh_token`), so it is easier for us to validate new features. It will be very useful, I believe.

File: source/extensions/filters/http/oauth2/filter.cc
Comment (Alexcei88) [reply source/extensions/filters/http/oauth2/filter.cc:515]: I agree, your name is better

File: source/extensions/filters/http/oauth2/filter.cc
Comment (Alexcei88) [reply source/extensions/filters/http/oauth2/filter.cc:376]: Because response_headers_to_add_ is not null only in the case of refreshing token. In other cases it is null

**Comment by Alexcei88** [comment]
> Thank you @Alexcei88. Flushing some comments. Also probably we need an example of using this filter under the: `examples` directory (this will be in a separate PR though, seems like you have it on one of your branches: `oauth2_refresh_token`), so it is easier for us to validate new features. It will be very useful, I believe.

I hope I will have time and mae an examle

File: source/extensions/filters/http/oauth2/oauth_client.cc
Comment (Alexcei88) [reply source/extensions/filters/http/oauth2/oauth_client.cc:139]: No, the exception is thrown when the programmer will write a bad code. It doesn't depend from answer of oauth server.

File: changelogs/current.yaml
Comment (Alexcei88) [reply changelogs/current.yaml]: I have updated the file

**Comment by Alexcei88** [comment]
> Thanks, the overall flow seems right to me, but this needs a bit of work otherwise
> 
> I think this would benefit from having a test in oauth_integration_test that verifies that we correctly pause filter iteration while waiting for a response for the refresh request. There are also a few branches that need test coverage per the coverage report (for example in https://storage.googleapis.com/envoy-pr/e9fdd37/coverage/source/extensions/filters/http/oauth2/filter.cc.gcov.html )

Sorry, but I don't have time understand how to write integration test. If anyone can help me, I will appreciate helping. I wrote enough unit tests.

**Comment by yanavlasov** [comment]
Hi @Alexcei88 The integration test suite already exists. https://github.com/envoyproxy/envoy/blob/main/test/extensions/filters/http/oauth2/oauth_integration_test.cc It allows end to end testin of the entire system. You can look at the `doAuthenticationFlow` function for orchestrating the authentication flow.

**Comment by RyanTheOptimist** [comment]
/wait
Looks like this is still waiting on Integration Tests

**Comment by Alexcei88** [comment]
> /wait Looks like this is still waiting on Integration Tests

It's ready

**Comment by wbpcode** [comment]
I will take a look today or tomorrow. sorry for the delay.

File: source/extensions/filters/http/oauth2/filter.cc
Comment (wbpcode) [inline source/extensions/filters/http/oauth2/filter.cc:195]: `(!token_.empty() && !refresh_token_.empty())`

File: source/extensions/filters/http/oauth2/filter.cc
Comment (wbpcode) [inline source/extensions/filters/http/oauth2/filter.cc:576]: Why we need create a header map here? I think a simple bool is enough here to tell we should update the response headers?

**Review by wbpcode — CHANGES_REQUESTED** [review]
Thanks for your great contribution. This is a big PR and I am not a oauth expert. So only some minor coments are added first. Seems it's good overall.

There are some historical comments that not be addressed completely but be marked as resolved.
So, I am afraid there are some important concerns that should be resolved but accutally not. And it's hard to iterate all these comments.

File: source/extensions/filters/http/oauth2/filter.cc
Comment (Alexcei88) [reply source/extensions/filters/http/oauth2/filter.cc:576]: We need to pass a map object to addResponseCookies. So, we create object. addResponseCookies method populate headers to the map

Comment (wbpcode) [reply source/extensions/filters/http/oauth2/filter.cc:575]: I mean why cannot we just call the addResponseCookies in the `encodeHeaders()` and use the accutally response headers as the parameter. Then we needn't create a unnecessary temporary object here. did I miss something?

**Comment by wbpcode** [comment]
Thanks for your contribution. And please don't use force push, esp for a big PR like this. It make it's very hard to do a gradually review.

**Comment by Alexcei88** [comment]
@wbpcode all is ready

File: source/extensions/filters/http/oauth2/filter.cc
Comment (wbpcode) [inline source/extensions/filters/http/oauth2/filter.cc:543]: nit: use `finishRefreshAccessTokenFlow` because the `onRefreshAccessTokenSuccess`. I think it would be better to unify these name to use `Update` or `Refresh`.

File: source/extensions/filters/http/oauth2/filter.cc
Comment (wbpcode) [inline source/extensions/filters/http/oauth2/filter.cc:579]: nit: `onRefreshAccessTokenFailure`

File: source/extensions/filters/http/oauth2/filter.h
Comment (wbpcode) [inline source/extensions/filters/http/oauth2/filter.h:184]: nit: `const bool use_refresh_token_{};`

**Review by wbpcode — COMMENTED** [review]
LGTM overall to these code. Thanks for your contributions. And only some nit comments to the naming.

And friendly ping cc @markdroth for API review.

**Comment by wbpcode** [comment]
Seem @markdroth is too busy recently, re-assign this to @adisuissa for api review.

File: api/envoy/extensions/filters/http/oauth2/v3/oauth.proto
Comment (adisuissa) [reply api/envoy/extensions/filters/http/oauth2/v3/oauth.proto:129]: I see your point of not breaking current behavior, and I agree it should be false by default at the moment.
It may be better to use a BoolValue type, as in the future it may make sense to convert the default value to true (see https://github.com/envoyproxy/envoy/blob/main/api/STYLE.md#api-style-guidelines re "wrapped scalar types").

**Review by adisuissa — COMMENTED** [review]
Thanks!
Left a minor comment re API.

File: api/envoy/extensions/filters/http/oauth2/v3/oauth.proto
Comment (Alexcei88) [reply api/envoy/extensions/filters/http/oauth2/v3/oauth.proto:129]: I don't want to make this changes, because I don't have more time to this PR. I killed very much time to fix minor changes. Sorry! I think it is easy to setup exactly refresh token support.

**Comment by Alexcei88** [comment]
What are we waiting yet?

File: api/envoy/extensions/filters/http/oauth2/v3/oauth.proto
Comment (adisuissa) [reply api/envoy/extensions/filters/http/oauth2/v3/oauth.proto:129]: Can you please elaborate what you mean by "it is easy to exactly set a refresh token support"?

**Comment by wbpcode** [comment]
I think approvement from @envoyproxy/api-shepherds is necessary. cc @adisuissa any concern?

/wait-any

**Comment by adisuissa** [comment]
> I think approvement from @envoyproxy/api-shepherds is necessary. cc @adisuissa any concern?
> 
> /wait-any

Might be hard to track in the github UI, but [this](https://github.com/envoyproxy/envoy/pull/24684#discussion_r1259133344) was reopened.

/wait-any

File: source/extensions/filters/http/oauth2/filter.cc
Comment (Alexcei88) [reply source/extensions/filters/http/oauth2/filter.cc:551]: @snowp , where do you see I touch other cookies except only the ones that are related with oauth2? Take a look please `finishRefreshAccessTokenFlow` method.
I only update `cookies oauth_hmac, oauth_expires, bearer_token, id_token, refresh_token` cookies.

File: api/envoy/extensions/filters/http/oauth2/v3/oauth.proto
Comment (Alexcei88) [reply api/envoy/extensions/filters/http/oauth2/v3/oauth.proto:129]: I mean user can set true or false clearly on configuration if user knows whether or not the oauth server supports refresh token.

Comment (adisuissa) [reply api/envoy/extensions/filters/http/oauth2/v3/oauth.proto:129]: > The default is false. Because. I don't consider to change automatically old behaviour after update envoy. If user wants to use refresh token then manually set the flag to true. In the future we can change default behaviour

If in the future we want to be able to update the default value of this field, it should be a wrapped type, rather than a primitive.
/wait-any

**Review by derekargueta — COMMENTED** [review]
some minor doc suggestions

**Comment by phlax** [comment]
@adisuissa @wbpcode i think this is still waiting on signoff

/wait-any

**Comment by GuillermoGarciaF** [comment]
Hi, could you confirm if the functionality will be merged and made available?
It is an important feature for many of us. Thank you!

**Comment by Alexcei88** [comment]
Hi @GuillermoGarciaF , No the code is not merged to main.

I have been supporting the PR for half of year.  It is very expensive for time. 
The review of PR is very slow.  Mainly due to some small changes. 
This functionality is working excelllent in production server for half of year where I work. But we cannot update Envoy without this PR. And I am interseting to finish this PR.

But periodically the new changes from main branch force me to rewrite new tests for oauth2 filter. 
Last I faced this problem again. And I don't have working environment to make fixes, because I am coding mainly on C#.
So I decided to close MR request and I will return when I will have more time.  
Sorry guys, but I will back sure a little bit later.

**Comment by roeap** [comment]
Thanks for your hard work @Alexcei88!

hoping that this eventually makes it to main, as it would be a great feature for our deployments as well.

File: api/envoy/extensions/filters/http/oauth2/v3/oauth.proto
Comment (Alexcei88) [reply api/envoy/extensions/filters/http/oauth2/v3/oauth.proto:129]: I have changed type of veriable to BoolValue as you requested

**Comment by wbpcode** [comment]
cc @Alexcei88 Thanks very much for your time and patience. And sorry for all bad experiences in the review processing.

I think the last thing need to be addressed is the API after my last review.

File: api/envoy/extensions/filters/http/oauth2/v3/oauth.proto
Comment (wbpcode) [inline api/envoy/extensions/filters/http/oauth2/v3/oauth.proto:130]: nit: add the end `.` after this sentence. This will be used to generate docs, so the formatter is also important.

File: api/envoy/extensions/filters/http/oauth2/v3/oauth.proto
Comment (adisuissa) [inline api/envoy/extensions/filters/http/oauth2/v3/oauth.proto:129]:
```suggestion
  // If set to true, allows automatic access token refresh using the associated refresh token (see 
  // `RFC 6749 section 6<https://datatracker.ietf.org/doc/html/rfc6749#section-6>`_), provided that the OAuth server supports that.
  // Default value is false.
```

**Review by adisuissa — COMMENTED** [review]
Thanks!
Left a small suggestion on the API description, but otherwise LGTM.
Will let @derekargueta have a final say on the implementation, as code-owner.

**Review by adisuissa — COMMENTED** [review]
/lgtm api

**Comment by Alexcei88** [comment]
@adisuissa , your last suggestion broke the [build](https://dev.azure.com/cncf/envoy/_build/results?buildId=150665&view=logs&j=9fd58ae0-0c50-56b1-f517-c2b9afda9c9b&t=c8fc762c-55bb-57d4-4cab-8ec527e7a423).

> ERROR: Unknown target name: "rfc 6749 section 6<[https://datatracker.ietf.org/doc/html/rfc6749#section-6>".](https://datatracker.ietf.org/doc/html/rfc6749#section-6%3E%22.)


Could you suggest how to fix it?

File: api/envoy/extensions/filters/http/oauth2/v3/oauth.proto
Comment (zuercher) [inline api/envoy/extensions/filters/http/oauth2/v3/oauth.proto:130]: I believe this needs a space after `section 6` and before the angle bracket.

See https://sublime-and-sphinx-guide.readthedocs.io/en/latest/references.html#links-to-external-web-pages

**Comment by wbpcode** [comment]
friendly ping @derekargueta

**Comment by wbpcode** [comment]
Seems @derekargueta is inactive recently. I am inclined to merge this after the format problem is resolved. cc @adisuissa

File: api/envoy/extensions/filters/http/oauth2/v3/oauth.proto
Comment (Alexcei88) [reply api/envoy/extensions/filters/http/oauth2/v3/oauth.proto]: @zuercher , Thank you, I will try

**Comment by adisuissa** [comment]
> Seems @derekargueta is inactive recently. I am inclined to merge this after the format problem is resolved. cc @adisuissa

@mattklein123 is the other code-owner, assigning Matt to give final review on this.
/assign @mattklein123

**Review by mattklein123 — APPROVED** [review]
Skimmed at a high level. I don't have the time/knowledge to understand the security implications of this.

**Review by wbpcode — APPROVED** [review]

**Pull request merged**
